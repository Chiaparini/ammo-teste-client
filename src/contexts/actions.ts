interface FetchProductsInit {
    type: 'FETCH_PRODUCTS_INIT'
};

interface FetchProductsSuccess {
    payload: any,
    type: 'FETCH_PRODUCTS_SUCCESS'
};

interface FetchProductsFail {
    type: 'FETCH_PRODUCTS_FAIL'
};

interface SetQuery {
    payload: any,
    type: 'SET_QUERY'
};

interface SetPageOrLimit {
    payload: any,
    type: 'SET_PAGE_OR_LIMIT'
};

interface SetTitle {
    payload: string,
    type: 'SET_TITLE'
};

interface ResetParams {
    type: 'RESET_PARAMS'
};


export type Actions = FetchProductsInit | FetchProductsSuccess | FetchProductsFail | SetQuery | SetPageOrLimit | SetTitle | ResetParams;