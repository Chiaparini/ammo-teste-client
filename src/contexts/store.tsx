import React, { useReducer, useEffect } from 'react';
import { Catalog } from '../models/catalog';
import { Actions } from './actions';
import { fetchProducts } from '../services/catalog';

export const CatalogContext = React.createContext({});

const initialState: Catalog = {
  loading: false,
  products: [],
  error: false,
  pages: 1,
  query: '',
  title: 'Lista de Produtos',
  page: 1,
  limit: 8,
  total: 0
};

const catalogReducer = (state: Catalog, action: Actions): any => {
  switch (action.type) {
    case 'FETCH_PRODUCTS_INIT':
      return { ...state, loading: true };
    case 'FETCH_PRODUCTS_SUCCESS':
      return { ...state, products: action.payload.products, pages: action.payload.pages, total: action.payload.total, loading: false, error: false };
    case 'FETCH_PRODUCTS_FAIL':
      return { ...state, error: true, loading: false };
    case 'SET_QUERY':
      return { ...state, query: action.payload.query };
    case 'SET_PAGE_OR_LIMIT':
      return { ...state, limit: action.payload.limit, page: action.payload.page };
    case 'SET_TITLE':
      return { ...state, title: action.payload }
    case 'RESET_PARAMS':
      return { ...initialState };
    default:
      throw new Error();
  }
};

export const useCatalog = () => {
  const [state, dispatch] = useReducer(catalogReducer, initialState);
  const currentUrlParams = new URLSearchParams(window.location.search);

  // get params when page renders 
  useEffect(() => {
    state.page = Number(currentUrlParams.get('page')) || initialState.page;
    state.query = Number(currentUrlParams.get('search')) || initialState.query;
    state.limit = Number(currentUrlParams.get('limit')) || initialState.limit;
  }, []);
  
  useEffect(() => {
    getProducts();

    return () => {};
  }, [state.page, state.limit]);


  const getProducts = async () => {
    dispatch({ type: 'FETCH_PRODUCTS_INIT' });
    try {
      currentUrlParams.set('search', state.query);
      currentUrlParams.set('limit', state.limit);
      currentUrlParams.set('page', state.page);

      // update URL params
      window.history.pushState({page: state.page, search: state.search, limit: state.limit}, '', '?'+currentUrlParams.toString());

      const products = await fetchProducts(state.limit, state.query, state.page);

      dispatch({ type: 'SET_TITLE', payload: state.query });
      dispatch({ type: 'FETCH_PRODUCTS_SUCCESS', payload: products });
    } catch {
      dispatch({ type: 'FETCH_PRODUCTS_FAIL' });
    }
  };

  const setQuery = (query: string) => {
    try {
      dispatch({ type: 'SET_QUERY', payload: { query } });
    } catch {
      dispatch({ type: 'RESET_PARAMS' });
    }
  };

  const setSensitiveParams = (page: number, limit: number) => {
    try {
      dispatch({ type: 'SET_PAGE_OR_LIMIT', payload: { page, limit } });
    } catch {
      dispatch({ type: 'RESET_PARAMS' });
    }
  };

  return [state, { getProducts, setQuery, setSensitiveParams }];
}

export const StoreProvider = ({ children }: any) => {
  const [state, action] = useCatalog();

  return (
    <CatalogContext.Provider value={{ state, action }}>
      {children}
    </CatalogContext.Provider>
  )
};