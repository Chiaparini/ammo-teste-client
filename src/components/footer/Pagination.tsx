import React, { useContext, useState, useEffect } from 'react'
import { CatalogContext } from '../../contexts/store';
import styled, { css } from 'styled-components';
import { colors, sizes } from '../../assets/style/global';

// Icons
import { ReactComponent as FirstPage } from '../../assets/images/first-page.svg';
import { ReactComponent as LastPage } from '../../assets/images/last-page.svg';
import { ReactComponent as PreviousPage } from '../../assets/images/chevron-left.svg';
import { ReactComponent as NextPage } from '../../assets/images/chevron-right.svg';

const PaginationWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;

  order: 1;

  margin-bottom: 22px;

  @media (min-width: ${sizes.md}) {
    margin: 0;
    order: 2;
  }
`;

const Button = styled.a<{ disabled: boolean, active?: boolean }>`
  display: flex;
  align-items: center;

  padding: 4px;
  border-radius: 5px;
  margin-right: 4px;
  border: 1px solid transparent;
  cursor: pointer;
  user-select: none;
  
  font-size: 12px;
  line-height: 32px;
  text-align: center;
  
  height: 32px;
  min-width: 22px;

  &:last-child {
    margin-right: 0;
  }

  svg {
    width: 16px;
    height: 16px;
    
    @media (min-width: ${sizes.md}) {
      width: 22px;
      height: 22px;
    }
  }

  @media (min-width: ${sizes.md}) {
    height: 32px;
    font-size: 16px;
    line-height: 22px;
  }

  ${(props: any) => props.disabled && css`
    pointer-events: none;
    cursor: normal;

    svg {
      fill: ${colors.lightGrey};
    }
  `}

  ${(props: any) => props.active && css`
    border: 1px solid ${colors.lightGrey};
    padding: 4px 16px;
  `}
`;

export const Pagination = () => {
  const {state, action}: any = useContext(CatalogContext);

  // state to control the number of showing pages
  const [pager, setPager] = useState({startPage: 0, endPage: 0});
  
  useEffect(() => {
    // set the maximum pages
    if (state.pages <= 6) {
      setPager({
        startPage: 1,
        endPage: state.pages
      });
    } else {
      if (state.page <= 3) {
        setPager({
          startPage: 1,
          endPage: 6
        });
      } else if (state.page + 2 >= state.pages) {
        setPager({
          startPage: state.pages - 5,
          endPage: state.pages
        });
      } else {
        setPager({
          startPage: state.page - 2,
          endPage: state.page + 3
        });
      }
    }
    return () => {};
  }, [state.pages, state.page])


  const goToFirstPage = () => {
    action.setSensitiveParams(1, state.limit);
  }

  const goToLastPage = () => {
    const lastpage = state.pages;
    action.setSensitiveParams(lastpage, state.limit);
  }

  const goToNextPage = () => {
    action.setSensitiveParams(state.page += 1, state.limit);
  }

  const goToPreviousPage = () => {
    action.setSensitiveParams(state.page -= 1, state.limit);
  }

  const shouldRenderPageButton = (currentIndex: number) => {
    if (currentIndex >= pager.startPage && currentIndex <= pager.endPage) {
      return (
        <Button 
            key={currentIndex}
            active={(Number(state.page) === currentIndex)} 
            disabled={(Number(state.page) === currentIndex)} 
            onClick={() => action.setSensitiveParams(currentIndex, state.limit) }>
              {currentIndex}
        </Button>
      )
    }
  }

  if (state.pages === 0) return null;

  return (
    <PaginationWrapper>
      <Button disabled={(state.page === 1)} onClick={goToFirstPage}>
        <FirstPage />
      </Button>
      <Button disabled={(state.page === 1)} onClick={goToPreviousPage}>
        <PreviousPage />
      </Button>
      {[...Array(state.pages)].map((v, index) =>
        shouldRenderPageButton(index + 1)
      )}
      <Button disabled={(state.page === state.pages)} onClick={goToNextPage}>
        <NextPage />
      </Button>
      <Button disabled={(state.page === state.pages)} onClick={goToLastPage}>
        <LastPage />
      </Button>
    </PaginationWrapper>
  )
}