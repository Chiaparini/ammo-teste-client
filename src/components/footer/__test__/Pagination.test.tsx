import React from 'react';
import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Pagination } from '../Pagination';
import { StoreProvider } from '../../../contexts/store';

afterEach(cleanup);

describe('Pagination', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <StoreProvider>
          <Pagination></Pagination>
        </StoreProvider>
        )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});