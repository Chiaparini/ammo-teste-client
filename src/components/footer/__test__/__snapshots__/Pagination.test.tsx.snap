// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<div
  className="sc-AykKC eTDbyt"
>
  <a
    className="sc-AykKD QbgV"
    disabled={true}
    onClick={[Function]}
  >
    <svg>
      first-page.svg
    </svg>
  </a>
  <a
    className="sc-AykKD QbgV"
    disabled={true}
    onClick={[Function]}
  >
    <svg>
      chevron-left.svg
    </svg>
  </a>
  <a
    className="sc-AykKD QbgV"
    disabled={true}
    onClick={[Function]}
  >
    <svg>
      chevron-right.svg
    </svg>
  </a>
  <a
    className="sc-AykKD QbgV"
    disabled={true}
    onClick={[Function]}
  >
    <svg>
      last-page.svg
    </svg>
  </a>
</div>
`;
